apps/mobile-wallet/src/features/modals/useBottomModalState.tsx

```diff
@@ -0,0 +1,260 @@
+    'worklet'
+
+    navHeight.value = withSpring(0, springConfig)
+    modalHeight.value = withSpring(0, springConfig, (finished) => finished && runOnJS(handleCloseOnJS)())
+    position.value = 'closing'
+  }, [handleCloseOnJS, modalHeight, navHeight, position])
+
+  const handleMaximize = useCallback(() => {
+    'worklet'
+    navHeight.value = withSpring(customNavHeight, springConfig)
+    modalHeight.value = withSpring(-maxHeight, springConfig)
+    position.value = 'maximised'
+  }, [navHeight, customNavHeight, modalHeight, maxHeight, position])
+
+  const handleMinimize = useCallback(() => {
+    'worklet'
+    navHeight.value = withSpring(0, springConfig)
+    modalHeight.value = withSpring(-minHeight.value, springConfig)
+    position.value = 'minimised'
+  }, [navHeight, modalHeight, minHeight, position])
+
+  const handleContentSizeChange = useCallback(
+    (_w: number, newContentHeight: number) => {
+      if (!modalHeight.value || newContentHeight > contentHeight.value + 1) {
+        runOnUI(() => {
+          contentHeight.value = newContentHeight
+          canMaximize.value = contentHeight.value > maxHeight
+          shouldMaximizeOnOpen.value = maximisedContent || contentHeight.value > maxHeight
+
+          minHeight.value = customMinHeight
+            ? customMinHeight
+            : shouldMaximizeOnOpen.value
+              ? maxHeight
+              : contentHeight.value + customNavHeight + insets.bottom
+
+          shouldMaximizeOnOpen.value ? handleMaximize() : handleMinimize()
+
+          // Determine if scrolling is needed
+          runOnJS(setIsScrollable)(contentHeight.value > dimensions.height * 0.9)
+        })()
+      }
+    },
+    [
+      modalHeight.value,
+      contentHeight,
+      canMaximize,
+      maxHeight,
+      shouldMaximizeOnOpen,
+      maximisedContent,
+      minHeight,
+      customMinHeight,
+      customNavHeight,
+      insets.bottom,
+      handleMaximize,
+      handleMinimize,
+      dimensions.height
+    ]
+  )
+
+  // Content scroll handlers
+  const contentScrollHandlers: ContentScrollHandlers = {
+    onScroll: (e) => {
+      contentScrollY.value = e.nativeEvent.contentOffset.y
+
+      if (!isContentDragged.value) return
+
+      if (contentScrollY.value <= 0) {
+        // Move the modal
+        if (contentScrollY.value < previousContentScrollY.value) {
+          const newModalHeightValue = modalHeight.value - contentScrollY.value
+          modalHeightDelta.value = modalHeight.value - newModalHeightValue
+          modalHeight.value = newModalHeightValue
+        }
+      } else if (-modalHeight.value < maxHeight) {
+        handleMaximize()
+      }
+      previousContentScrollY.value = contentScrollY.value
+    },
+    onScrollBeginDrag: () => {
+      isContentDragged.value = true
+    },
+    onScrollEndDrag: () => {
+      isContentDragged.value = false
+
+      if (modalHeightDelta.value < -1) {
+        handleClose()
+      }
+    }
+  }
```

Comment:
So, that was a pain in the 🍑 
It's not perfect yet, there are some glitches when the use scrolls down the content a bit, then go back up again. But for closing modals quickly it works well.